src/System.Text.Json/src/System/Text/Json/JsonDocument.CustomDb.cs

```diff
@@ -0,0 +1,229 @@
+                ArrayPool<byte>.Shared.Return(_rentedBuffer);
+                _rentedBuffer = newArray;
+            }
+
+            [Conditional("DEBUG")]
+            private void AssertValidIndex(int index)
+            {
+                Debug.Assert(index >= 0);
+                Debug.Assert(index <= Length - DbRow.Size, $"index {index} is out of bounds");
+                Debug.Assert(index % DbRow.Size == 0, $"index {index} is not at a record start position");
+            }
+
+            internal void SetLength(int index, int length)
+            {
+                AssertValidIndex(index);
+                Debug.Assert(length >= 0);
+                Span<byte> destination = _rentedBuffer.AsSpan(index + SizeOrLengthOffset);
+                MemoryMarshal.Write(destination, ref length);
+            }
+
+            internal void SetNumberOfRows(int index, int numberOfRows)
+            {
+                AssertValidIndex(index);
+                Debug.Assert(numberOfRows >= 1 && numberOfRows <= 0x0FFFFFFF);
+
+                Span<byte> dataPos = _rentedBuffer.AsSpan(index + NumberOfRowsOffset);
+                int current = MemoryMarshal.Read<int>(dataPos);
+
+                // Persist the most significant nybble
+                int value = (current & unchecked((int)0xF0000000)) | numberOfRows;
+                MemoryMarshal.Write(dataPos, ref value);
+            }
+
+            internal void SetHasComplexChildren(int index)
+            {
+                AssertValidIndex(index);
+
+                // The HasComplexChildren bit is the most significant bit of "SizeOrLength"
+                Span<byte> dataPos = _rentedBuffer.AsSpan(index + SizeOrLengthOffset);
+                int current = MemoryMarshal.Read<int>(dataPos);
+
+                int value = current | unchecked((int)0x80000000);
+                MemoryMarshal.Write(dataPos, ref value);
+            }
+
+            internal int FindIndexOfFirstUnsetSizeOrLength(JsonTokenType lookupType)
+            {
+                Debug.Assert(lookupType == JsonTokenType.StartObject || lookupType == JsonTokenType.StartArray);
+                return FindOpenElement(lookupType);
+            }
+
+            private int FindOpenElement(JsonTokenType lookupType)
+            {
+                Span<byte> data = _rentedBuffer.AsSpan(0, Length);
+               
+                for (int i = Length - DbRow.Size; i >= 0; i -= DbRow.Size)
+                {
+                    DbRow row = MemoryMarshal.Read<DbRow>(data.Slice(i));
+
+                    if (row.IsUnknownSize && row.TokenType == lookupType)
+                    {
+                        return i;
+                    }
+                }
+
+                // We should never reach here.
+                Debug.Fail($"Unable to find expected {lookupType} token");
+                return -1;
+            }
+
+            internal void Get(int index, out DbRow row)
```

Comment:
Why an `out` rather than a return value?

Reply:
I think there used to be a chain, and I was going for "write once, read many"; but everything's a direct caller now, and it's only 12 bytes, so simple return seems simpler.